File: ios/ClaviKeyboard/ClaviKeyboardExtension/TranslitEngine.swift
import Foundation

/**
 * KMU 2010 Ukrainian transliteration engine.
 * Mirrors TranslitEngine.kt — longest-match-first.
 */
struct TranslitEngine {

    private struct Rule {
        let latin: String
        let cyrillic: String
    }

    private let rules: [Rule] = [
        // Trigraphs
        Rule(latin: "shch", cyrillic: "щ"), Rule(latin: "Shch", cyrillic: "Щ"), Rule(latin: "SHCH", cyrillic: "Щ"),
        // Digraphs
        Rule(latin: "zh", cyrillic: "ж"), Rule(latin: "Zh", cyrillic: "Ж"), Rule(latin: "ZH", cyrillic: "Ж"),
        Rule(latin: "kh", cyrillic: "х"), Rule(latin: "Kh", cyrillic: "Х"), Rule(latin: "KH", cyrillic: "Х"),
        Rule(latin: "ts", cyrillic: "ц"), Rule(latin: "Ts", cyrillic: "Ц"), Rule(latin: "TS", cyrillic: "Ц"),
        Rule(latin: "ch", cyrillic: "ч"), Rule(latin: "Ch", cyrillic: "Ч"), Rule(latin: "CH", cyrillic: "Ч"),
        Rule(latin: "sh", cyrillic: "ш"), Rule(latin: "Sh", cyrillic: "Ш"), Rule(latin: "SH", cyrillic: "Ш"),
        Rule(latin: "iu", cyrillic: "ю"), Rule(latin: "Iu", cyrillic: "Ю"), Rule(latin: "IU", cyrillic: "Ю"),
        Rule(latin: "ia", cyrillic: "я"), Rule(latin: "Ia", cyrillic: "Я"), Rule(latin: "IA", cyrillic: "Я"),
        Rule(latin: "ie", cyrillic: "є"), Rule(latin: "Ie", cyrillic: "Є"), Rule(latin: "IE", cyrillic: "Є"),
        Rule(latin: "yi", cyrillic: "ї"), Rule(latin: "Yi", cyrillic: "Ї"), Rule(latin: "YI", cyrillic: "Ї"),
        // Apostrophe
        Rule(latin: "'", cyrillic: "\u{02BC}"),
        // Single chars
        Rule(latin: "a", cyrillic: "а"), Rule(latin: "A", cyrillic: "А"),
        Rule(latin: "b", cyrillic: "б"), Rule(latin: "B", cyrillic: "Б"),
        Rule(latin: "v", cyrillic: "в"), Rule(latin: "V", cyrillic: "В"),
        Rule(latin: "h", cyrillic: "г"), Rule(latin: "H", cyrillic: "Г"),
        Rule(latin: "g", cyrillic: "ґ"), Rule(latin: "G", cyrillic: "Ґ"),
        Rule(latin: "d", cyrillic: "д"), Rule(latin: "D", cyrillic: "Д"),
        Rule(latin: "e", cyrillic: "е"), Rule(latin: "E", cyrillic: "Е"),
        Rule(latin: "z", cyrillic: "з"), Rule(latin: "Z", cyrillic: "З"),
        Rule(latin: "y", cyrillic: "и"), Rule(latin: "Y", cyrillic: "И"),
        Rule(latin: "i", cyrillic: "і"), Rule(latin: "I", cyrillic: "І"),
        Rule(latin: "j", cyrillic: "й"), Rule(latin: "J", cyrillic: "Й"),
        Rule(latin: "k", cyrillic: "к"), Rule(latin: "K", cyrillic: "К"),
        Rule(latin: "l", cyrillic: "л"), Rule(latin: "L", cyrillic: "Л"),
        Rule(latin: "m", cyrillic: "м"), Rule(latin: "M", cyrillic: "М"),
        Rule(latin: "n", cyrillic: "н"), Rule(latin: "N", cyrillic: "Н"),
        Rule(latin: "o", cyrillic: "о"), Rule(latin: "O", cyrillic: "О"),
        Rule(latin: "p", cyrillic: "п"), Rule(latin: "P", cyrillic: "П"),
        Rule(latin: "r", cyrillic: "р"), Rule(latin: "R", cyrillic: "Р"),
        Rule(latin: "s", cyrillic: "с"), Rule(latin: "S", cyrillic: "С"),
        Rule(latin: "t", cyrillic: "т"), Rule(latin: "T", cyrillic: "Т"),
        Rule(latin: "u", cyrillic: "у"), Rule(latin: "U", cyrillic: "У"),
        Rule(latin: "f", cyrillic: "ф"), Rule(latin: "F", cyrillic: "Ф"),
        Rule(latin: "x", cyrillic: "кс"), Rule(latin: "X", cyrillic: "КС"),
        Rule(latin: "c", cyrillic: "с"), Rule(latin: "C", cyrillic: "С"),
        Rule(latin: "q", cyrillic: "к"), Rule(latin: "Q", cyrillic: "К"),
        Rule(latin: "w", cyrillic: "в"), Rule(latin: "W", cyrillic: "В"),
    ]

    private var maxLen: Int { rules.map { $0.latin.count }.max() ?? 1 }

    /// Try to transliterate the beginning of `buffer`. Returns (result, charsConsumed) or nil.
    func tryTransliterate(_ buffer: String) -> (String, Int)? {
        let maxLen = self.maxLen
        for len in stride(from: min(maxLen, buffer.count), through: 1, by: -1) {
            let candidate = String(buffer.prefix(len))
            if let rule = rules.first(where: { $0.latin == candidate }) {
                return (rule.cyrillic, len)
            }
        }
        return nil
    }

    func isDigraphStart(_ c: Character) -> Bool {
        "szktciy".contains(c.lowercased())
    }
}

File: android/app/src/main/java/com/clavi/keyboard/TranslitEngine.kt
package com.clavi.keyboard

/**
 * KMU 2010 Ukrainian transliteration engine.
 * Latin (phonetic) -> Ukrainian Cyrillic.
 * Longest-match-first strategy for digraphs/trigraphs.
 */
class TranslitEngine {

    private data class Rule(val latin: String, val cyrillic: String)

    // Sorted by length descending so longest match wins
    private val rules: List<Rule> = listOf(
        // Trigraphs
        Rule("shch", "\u0449"), Rule("Shch", "\u0429"), Rule("SHCH", "\u0429"),
        // Digraphs
        Rule("zh", "\u0436"), Rule("Zh", "\u0416"), Rule("ZH", "\u0416"),
        Rule("kh", "\u0445"), Rule("Kh", "\u0425"), Rule("KH", "\u0425"),
        Rule("ts", "\u0446"), Rule("Ts", "\u0426"), Rule("TS", "\u0426"),
        Rule("ch", "\u0447"), Rule("Ch", "\u0427"), Rule("CH", "\u0427"),
        Rule("sh", "\u0448"), Rule("Sh", "\u0428"), Rule("SH", "\u0428"),
        // yu/ya — user-friendly variants (type ya→я, yu→ю)
        Rule("yu", "\u044E"), Rule("Yu", "\u042E"), Rule("YU", "\u042E"),
        Rule("ya", "\u044F"), Rule("Ya", "\u042F"), Rule("YA", "\u042F"),
        Rule("ye", "\u0454"), Rule("Ye", "\u0404"), Rule("YE", "\u0404"),
        Rule("yi", "\u0457"), Rule("Yi", "\u0407"), Rule("YI", "\u0407"),
        // Legacy ia/iu variants
        Rule("iu", "\u044E"), Rule("Iu", "\u042E"), Rule("IU", "\u042E"),
        Rule("ia", "\u044F"), Rule("Ia", "\u042F"), Rule("IA", "\u042F"),
        Rule("ie", "\u0454"), Rule("Ie", "\u0404"), Rule("IE", "\u0404"),
        // Apostrophe
        Rule("'", "\u02BC"),
        // Single chars
        Rule("a", "\u0430"), Rule("A", "\u0410"),
        Rule("b", "\u0431"), Rule("B", "\u0411"),
        Rule("v", "\u0432"), Rule("V", "\u0412"),
        Rule("h", "\u0433"), Rule("H", "\u0413"),
        Rule("g", "\u0491"), Rule("G", "\u0490"),
        Rule("d", "\u0434"), Rule("D", "\u0414"),
        Rule("e", "\u0435"), Rule("E", "\u0415"),
        Rule("z", "\u0437"), Rule("Z", "\u0417"),
        Rule("y", "\u0438"), Rule("Y", "\u0418"),
        Rule("i", "\u0456"), Rule("I", "\u0406"),
        Rule("j", "\u0439"), Rule("J", "\u0419"),
        Rule("k", "\u043A"), Rule("K", "\u041A"),
        Rule("l", "\u043B"), Rule("L", "\u041B"),
        Rule("m", "\u043C"), Rule("M", "\u041C"),
        Rule("n", "\u043D"), Rule("N", "\u041D"),
        Rule("o", "\u043E"), Rule("O", "\u041E"),
        Rule("p", "\u043F"), Rule("P", "\u041F"),
        Rule("r", "\u0440"), Rule("R", "\u0420"),
        Rule("s", "\u0441"), Rule("S", "\u0421"),
        Rule("t", "\u0442"), Rule("T", "\u0422"),
        Rule("u", "\u0443"), Rule("U", "\u0423"),
        Rule("f", "\u0444"), Rule("F", "\u0424"),
        Rule("x", "\u043A\u0441"), Rule("X", "\u041A\u0421"),
        Rule("c", "\u0441"), Rule("C", "\u0421"),
        Rule("q", "\u043A"), Rule("Q", "\u041A"),
        Rule("w", "\u0432"), Rule("W", "\u0412"),
    )

    private val maxRuleLen = rules.maxOf { it.latin.length }

    /**
     * Transliterate a full string from Latin to Ukrainian Cyrillic.
     */
    fun transliterate(input: String): String {
        val result = StringBuilder(input.length)
        var i = 0
        while (i < input.length) {
            var matched = false
            // Try longest match first
            val maxLen = minOf(maxRuleLen, input.length - i)
            for (len in maxLen downTo 1) {
                val candidate = input.substring(i, i + len)
                val rule = rules.find { it.latin == candidate }
                if (rule != null) {
                    result.append(rule.cyrillic)
                    i += len
                    matched = true
                    break
                }
            }
            if (!matched) {
                result.append(input[i])
                i++
            }
        }
        return result.toString()
    }

    /**
     * Transliterate a single character (for real-time input).
     * Returns null if the char might be part of a digraph and we need to buffer.
     */
    fun isDigraphStart(c: Char): Boolean {
        val lower = c.lowercaseChar()
        return lower in setOf('s', 'z', 'k', 't', 'c', 'i', 'y')
    }

    /**
     * Try to transliterate a buffered sequence. Returns the result and how many
     * chars were consumed, or null if no match.
     */
    fun tryTransliterate(buffer: String): Pair<String, Int>? {
        for (len in minOf(maxRuleLen, buffer.length) downTo 1) {
            val candidate = buffer.substring(0, len)
            val rule = rules.find { it.latin == candidate }
            if (rule != null) {
                return Pair(rule.cyrillic, len)
            }
        }
        return null
    }
}
